docs: add templated policies workflow example to kubernetes auth (#25694)
hashicorp/vault: website/content/docs/auth/kubernetes.mdx

[k8s-tokenreview]: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#tokenreview-v1-authentication-k8s-io

## Workflows

Refer to the following workflow examples for Kubernetes auth method usage:

### Working with templated policies

When configuring Kubernetes auth, you can set `use_annotations_as_alias_metadata=true`
to use Kubernetes Service Account annotations for
[Vault alias](/vault/docs/concepts/identity#entities-and-aliases) metadata. This will
allow you to utilize the `identity.entity.aliases.<mount accessor>.metadata.<metadata key>`
templating parameter in creating
[templated policies](/vault/docs/concepts/policies#templated-policies). Note that
Vault will need permission to read service accounts from the Kubernetes API.

#### Scenario Introduction

> Assume that the following policy requirement was given:
> 
> Each application can perform read operations on its allocated key/value secret path
> `(env-kv/data/<env>)`

Reviewer: ```suggestion

Assume you have the following policy requirement:

Applications can perform read operations on its allocated key/value secret path:
`(env-kv/data/<env>)`
```

Style correction: write in active voice

#### Annotate Kubernetes Service Accounts with their dedicated secret paths

```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: app
  namespace: demo
  annotations:
    vault.hashicorp.com/alias-metadata-env: demo/app
```

> The annotation will be rendered to the alias metadata as `env : demo/app`
> upon Vault login using `app` ServiceAccount's JWT.

Reviewer: ```suggestion
When the application, `app`, logs in with JWT for the service account, Vault
renders  the alias metadata as `env : demo/app`
```

#### Create a templated ACL policy

> The `env-tmpl` policy allows applications to read their secret that
> are defined in KV v2 secret engine. The mount accessor value (auth_kubernetes_bcecb1e1
> in this example) can be read from the sys/auth endpoint.

Reviewer: ```suggestion
The `env-tmpl` policy lets applications read their secrets defined in KV v2
secret engine. So you can read the example mount accessor value
(`auth_kubernetes_bcecb1e1`) from the `sys/auth` endpoint.
```

```bash
tee env-tmpl.hcl <<EOF
path "env-kv/data/{{identity.entity.aliases.auth_kubernetes_bcecb1e1.metadata.env}}" {
  capabilities = [ "read" ]
}
EOF

vault policy write env-tmpl env-tmpl.hcl
```

#### Create a Kubernetes role with the templated ACL policy

> This allows users that login as the `env-reader` role to read at the secret path
> described in the `env-tmpl` policy.

Reviewer: ```suggestion
The Kubernetes role lets users login as the `env-reader` role to read from the
secret path described in the `env-tmpl` policy.
```

configuration to use Kubernetes Service Account annotations for
[Vault alias](/vault/docs/concepts/identity#entities-and-aliases) metadata.

When `use_annotations_as_alias_metadata` is true, you can use the
`identity.entity.aliases.<mount accessor>.metadata.<metadata key>` template
parameter when you create [templated policies](/vault/docs/concepts/policies#templated-policies).

To use annotations as alias metadata, you must give Vault permission to read
service accounts from the Kubernetes API.

#### Scenario Introduction

Assume you have the following policy requirement:

Applications can perform read operations on its allocated key/value secret path:
`(env-kv/data/<env>)`

#### Annotate Kubernetes Service Accounts with their dedicated secret paths

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: app
  namespace: demo
  annotations:
    vault.hashicorp.com/alias-metadata-env: demo/app
```

When the application, `app`, logs in with JWT for the service account, Vault
renders the alias metadata as `env : demo/app`.

#### Create a templated ACL policy

The `env-tmpl` policy lets applications read their secrets defined in KV v2

> secret engine. So you can read the example mount accessor value
> (`auth_kubernetes_bcecb1e1`) from the `sys/auth` endpoint.

Reviewer 2: Just thinking this could be a little more direct, and link to sys/auth and "vault auth list".

```suggestion
secret engine. Use the mount accessor value
(`auth_kubernetes_bcecb1e1`) from the [`sys/auth`](https://developer.hashicorp.com/vault/api-docs/system/auth) endpoint or the [`vault auth list`](https://developer.hashicorp.com/vault/docs/commands/auth/list) command.
```

Reviewer 2: And oops, those should be relative links (`/vault/api-docs/system/auth`, `/vault/docs/commands/auth/list`)
